File: app/src/main/java/com/slapshotapps/dragonshockey/models/Game.java
package com.slapshotapps.dragonshockey.models;

import android.os.Parcel;
import android.os.Parcelable;
import android.support.annotation.Keep;
import com.google.firebase.database.Exclude;
import com.slapshotapps.dragonshockey.Utils.DateFormaters;
import java.util.Date;

/**
 * A custom object for a game
 */
@Keep
public class Game implements Parcelable, Cloneable {
    public String gameTime;
    public String opponent;
    public int gameID;
    public boolean home;

    @Exclude
    public GameResult gameResult;

    public Date gameTimeToDate() {
        if (gameTime == null) {
            return null;
        }

        return DateFormaters.getDateFromGameTime(gameTime);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        Game game = (Game) o;

        if (gameID != game.gameID) {
            return false;
        }
        if (gameTime != null ? !gameTime.equals(game.gameTime) : game.gameTime != null) {
            return false;
        }
        if (opponent != null ? !opponent.equals(game.opponent) : game.opponent != null) {
            return false;
        }
        return gameResult != null ? gameResult.equals(game.gameResult) : game.gameResult == null;
    }

    @Override
    public Game clone() throws CloneNotSupportedException {
        Game cloneGame = new Game();
        cloneGame.gameTime = gameTime;
        cloneGame.opponent = opponent;
        cloneGame.gameID = gameID;
        cloneGame.home = home;
        if (gameResult != null) {
            cloneGame.gameResult = new GameResult();
            cloneGame.gameResult.dragonsScore = gameResult.dragonsScore;
            cloneGame.gameResult.opponentScore = gameResult.opponentScore;
            cloneGame.gameResult.overtimeLoss = gameResult.overtimeLoss;
            cloneGame.gameResult.gameID = gameResult.gameID;
        }
        return cloneGame;
    }

    @Override
    public int hashCode() {
        int result = gameTime != null ? gameTime.hashCode() : 0;
        result = 31 * result + (opponent != null ? opponent.hashCode() : 0);
        result = 31 * result + gameID;
        result = 31 * result + (gameResult != null ? gameResult.hashCode() : 0);
        result = 31 * result + (home ? 1 : 0);
        return result;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeString(this.gameTime);
        dest.writeString(this.opponent);
        dest.writeInt(this.gameID);
        dest.writeParcelable(this.gameResult, flags);
        dest.writeByte((byte) (home ? 1 : 0));
    }

    public Game() {
    }

    protected Game(Parcel in) {
        this.gameTime = in.readString();
        this.opponent = in.readString();
        this.gameID = in.readInt();
        this.gameResult = in.readParcelable(GameResult.class.getClassLoader());
        this.home = in.readByte() == 1;
    }

    public static final Parcelable.Creator<Game> CREATOR = new Parcelable.Creator<Game>() {
        @Override
        public Game createFromParcel(Parcel source) {
            return new Game(source);
        }

        @Override
        public Game[] newArray(int size) {
            return new Game[size];
        }
    };
}

File: app/src/main/java/com/slapshotapps/dragonshockey/models/Game.kt
package com.slapshotapps.dragonshockey.models

import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.Keep
import com.google.firebase.database.Exclude
import com.slapshotapps.dragonshockey.Utils.DateFormatter
import java.util.*

/**
 * A custom object for a game
 */
@Keep
class Game : Cloneable, Parcelable {
    var gameTime: String? = null

    var opponent: String? = null

    var gameID: Int = 0

    var home: Boolean = false

    var rink: String? = null

    @Exclude
    var gameResult: GameResult? = null

    fun getRinkName(): String {

        return when(rink)
        {
            "east" -> "East Rink"
            "west" -> "West Rink"
            else -> ""
        }
    }

    fun gameTimeToDate(): Date? {
        val time = gameTime
        return if (time == null) {
            null
        } else DateFormatter.getDateFromGameTime(time)

    }

    override fun equals(other: Any?): Boolean {
        if (this === other) {
            return true
        }
        if (other == null || !(other is Game)) {
            return false
        }

        val game = other as Game?

        if (gameID != game!!.gameID) {
            return false
        }
        if (if (gameTime != null) gameTime != game.gameTime else game.gameTime != null) {
            return false
        }
        if (if (opponent != null) opponent != game.opponent else game.opponent != null) {
            return false
        }
        return if (gameResult != null) gameResult == game.gameResult else game.gameResult == null
    }

    @Throws(CloneNotSupportedException::class)
    public override fun clone(): Game {
        val cloneGame = Game()
        cloneGame.gameTime = gameTime
        cloneGame.opponent = opponent
        cloneGame.gameID = gameID
        cloneGame.home = home
        if (gameResult != null) {
            cloneGame.gameResult = GameResult()
            cloneGame.gameResult!!.dragonsScore = gameResult!!.dragonsScore
            cloneGame.gameResult!!.opponentScore = gameResult!!.opponentScore
            cloneGame.gameResult!!.overtimeLoss = gameResult!!.overtimeLoss
            cloneGame.gameResult!!.gameID = gameResult!!.gameID
        }
        return cloneGame
    }

    override fun hashCode(): Int {
        var result = if (gameTime != null) gameTime!!.hashCode() else 0
        result = 31 * result + if (opponent != null) opponent!!.hashCode() else 0
        result = 31 * result + gameID
        result = 31 * result + if (gameResult != null) gameResult!!.hashCode() else 0
        result = 31 * result + if (home) 1 else 0
        result = 31 * result + if (rink != null) rink!!.hashCode() else 0
        return result
    }

    constructor()

    constructor(source: Parcel) : this() {
        gameTime = source.readString()
        opponent = source.readString()
        gameID = source.readInt()
        gameResult = source.readParcelable(GameResult::class.java.classLoader)
        home = source.readByte().toInt() == 1
        rink = source.readString()
    }

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = with(dest) {
        writeString(gameTime)
        writeString(opponent)
        writeInt(gameID)
        writeParcelable(gameResult, flags)
        writeByte(if (home) 1 else 0)
        writeString(rink)
    }

    companion object {
        @JvmField
        val CREATOR: Parcelable.Creator<Game> = object : Parcelable.Creator<Game> {
            override fun createFromParcel(source: Parcel): Game = Game(source)
            override fun newArray(size: Int): Array<Game?> = arrayOfNulls(size)
        }
    }
}
